packages/outbox-core/lib/outbox.ts

```diff
@@ -0,0 +1,127 @@
+import { AbstractPeriodicJob, type JobExecutionContext } from '@lokalise/background-jobs-common'
+import type { PeriodicJobDependencies } from '@lokalise/background-jobs-common'
+import type {
+  CommonEventDefinition,
+  CommonEventDefinitionPublisherSchemaType,
+  ConsumerMessageMetadataType,
+  DomainEventEmitter,
+} from '@message-queue-toolkit/core'
+import { PromisePool } from '@supercharge/promise-pool'
+import { uuidv7 } from 'uuidv7'
+import type { OutboxAccumulator } from './accumulators'
+import type { OutboxStorage } from './storage'
+
+/**
+ * Main logic for handling outbox entries.
+ *
+ * If entry is rejected, it is NOT going to be handled during the same execution. Next execution will pick it up.
+ */
+export class OutboxProcessor<SupportedEvents extends CommonEventDefinition[]> {
+  constructor(
+    private readonly outboxStorage: OutboxStorage<SupportedEvents>,
+    private readonly outboxAccumulator: OutboxAccumulator<SupportedEvents>,
+    private readonly eventEmitter: DomainEventEmitter<SupportedEvents>,
+    private readonly maxRetryCount: number,
+    private readonly emitBatchSize: number,
+  ) {}
+
+  public async processOutboxEntries(context: JobExecutionContext) {
+    const entries = await this.outboxStorage.getEntries(this.maxRetryCount)
```

Comment:
can we also check the accumulator in the beginning to see if there is anything already processed there, and filter these out?

Reply:
Maybe instead let's clear before processing?

Reply:
But it's duplicates concern, right?

Maybe instead we should make sure that accumulator takes care of duplicated values? I.e. each outbox entry has its own unique id, accumulator could check if it already has give entry.

Reply:
I'm primarily concerned about Redis use-case here. If we use a set for storing elements, we could avoid persisting it to DB twice, but it wouldn't help us to avoid emitting the event twice.

What are the drawbacks of explicit prefiltering?

Reply:
Done.